src/rosdiscover/observer/observer.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+__all__ = ("Observer",)
+import contextlib
+from abc import ABC, abstractmethod
+from typing import Dict, Iterator
+import roswire
+from roswire import AppInstance, ROSVersion
+
+from ..config import Config
+from ..interpreter import NodeContext, SystemSummary
+
+
+class Observer(ABC):
+
+    @classmethod
+    @contextlib.contextmanager
+    def for_container(cls,
+                      container: str,
+                      config: Config,
+                      ) -> Iterator['Observer']:
+        """Constructs and interpreter for a given running container"""
+        rsw = roswire.ROSWire()
+        app = rsw.app(config.image, config.sources)
+        instance = app.attach(container, require_description=True)
+        if app.description.distribution.ros == ROSVersion.ROS1:
+            from .ros1 import ROS1Observer
+            yield ROS1Observer(instance, config)
+        else:
+            from .ros2 import ROS2Observer
+            yield ROS2Observer(instance, config)
+
+    def __init__(self, app: AppInstance, config: Config):
```

Comment:
An `ABC` with a constructor should be avoided. ABCs should ideally be used to specify an interface rather than to provide an implementation.

Reply:
I don't really agree with this, because this means would would need to either duplicate this code in each subclass or put it in some builder. Instead, I've made the `__init__` method abstract, as per: https://stackoverflow.com/questions/44800659/how-define-constructor-implementation-for-an-abstract-class-in-python